### PR TITLE
Periodic neighbors

### PR DESCRIPTION
**Pull request by masterleinad**
In this PR, the possibility to ask a cell for its periodic neighbors is implemented.
It is based on the discussion in
https://groups.google.com/forum/#!searchin/dealii/periodic$20neighbor/dealii/mCUUm-nqhoU/XG2XaJ2cFwAJ
Essentially all the information comes from periodic_face_map that stores both cell_indices, both face numbers and the relative orientation of these faces.  For multigrid support neighbors on all levels are stored. Furthermore, the functions in distributed/tria.cc related to multigrid now use periodic_face_map instead of periodic_face_pairs_level_0.

### CODE REVIEW

File: include/deal.II/grid/tria_accessor.h [hunk 1]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+   */
+  /**
+   * If the cell has a periodic neighbor at its @c ith face, this function
+   * returns true, otherwise, the returned value is false.
+   */
+  bool has_periodic_neighbor(const unsigned int i) const;
+
+  /**
+   * For a cell with its @c ith face at a periodic boundary,
+   * (see @ref GlossPeriodicConstraints "the entry for periodic baoundaries")
```
Comment (kronbichler) [inline include/deal.II/grid/tria_accessor.h]: baoundaries -> boundaries

File: include/deal.II/grid/tria_accessor.h [hunk 2]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+   * this function returns an iterator to the cell on the other side
+   * of the periodic boundary. If there is no periodic boundary at the @c ith
+   * face, an exception will be thrown.
+   * In order to check if a cell has periodic neighbor on its @c ith face,
+   * one should first call the function @c has_periodic_neighbor().
```
Comment (kronbichler) [inline include/deal.II/grid/tria_accessor.h]: Maybe reword: In order to avoid running into an exception, check the result of has_periodic_neighbor() for the ith face prior to this function.

File: include/deal.II/grid/tria_accessor.h [hunk 3]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+  periodic_neighbor (const unsigned int i) const;
+
+  /**
+   * Returns an iterator to the periodic neighbor of the cell at a given
+   * face and subface number. The general guidelines for using this function
+   * is similar to the function @c neighbor_child_on_subface. The
+   * implementation of this function is consistent with
+   * @c periodic_neighbor_of_coarser_periodic_neighbor. For instance,
+   * assume that we are sitting on a cell named @c cell1, which has a 1 level
+   * coarser neighbor at its @c ith face. Let us name this coarser neighbor
```
Comment (kronbichler) [inline include/deal.II/grid/tria_accessor.h]: "which has a 1 level coarser neighbor at its @c ith face" -> "whose neighbor behind the @c ith face is one level coarser"

File: include/deal.II/grid/tria_accessor.h [hunk 4]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+   * This function is a generalization of
+   * @c periodic_neighbor_of_periodic_neighbor
+   * for those cells which have a coarser periodic neighbor. The returned
+   * pair of numbers can be used in @c periodic_neighbor_child_on_subface
+   * to get back to the current cell. In other words, the following
+   * assertion should be true, for a cell with coarser periodic neighbor:
+   * cell->periodic_neighbor(i)->periodic_neighbor_child_on_subface(face_no, subface_no)==cell
+   */
+  std::pair<unsigned int, unsigned int>
+  periodic_neighbor_of_coarser_periodic_neighbor (const unsigned i) const;
```
Comment (kronbichler) [inline include/deal.II/grid/tria_accessor.h]: I assume you put the face number here. Please change the variable name to 'face_no' or similar.

File: include/deal.II/grid/tria_accessor.h [hunk 5]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+  /**
+   * This function returns the level of the periodic neighbor. If there is
+   * no periodic neighbor at the given face, the returned value is -1.
+   */
+  int
+  periodic_neighbor_level (const unsigned int i) const;
+
+  /**
+   * For a cell with a periodic neighbor at its @c ith face, this function
+   * returns the face number of that periodic neighbor such that, the
```
Comment (kronbichler) [inline include/deal.II/grid/tria_accessor.h]: remove comma

File: include/deal.II/grid/tria_accessor.h [hunk 6]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+  /**
+   * If a cell has periodic neighbor, this function returns the face number
+   * of the neighbor, which is connected to this cell.
+   */
+  unsigned int
+  periodic_neighbor_face_no (const unsigned int i) const;
+
+  /**
+   * This function returns true if the element on the other side of the
+   * periodic boundary is coarser and returns false, otherwise. The
```
Comment (kronbichler) [inline include/deal.II/grid/tria_accessor.h]: remove comma

File: include/deal.II/grid/tria_accessor.h [hunk 7]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+   * implementation allows this function to work, in the case of
```
Comment (kronbichler) [inline include/deal.II/grid/tria_accessor.h]: remove comma

File: source/distributed/tria.cc [hunk 8]
```diff
@@ -3124,92 +3124,14 @@ namespace parallel
...
-
-
-
       // ensures the 2:1 mesh balance for periodic boundary conditions in the
       // artificial cell layer (the active cells are taken care of by p4est)
       template <int dim, int spacedim>
       bool enforce_mesh_balance_over_periodic_boundaries
       (Triangulation<dim,spacedim> &tria,
-       const std::vector<GridTools::PeriodicFacePair<typename dealii::Triangulation<dim,spacedim>::cell_iterator> > periodic_face_pairs_level_0)
+       const std::vector<GridTools::PeriodicFacePair<typename dealii::Triangulation<dim,spacedim>::cell_iterator> > &periodic_face_pairs_level_0)
```
Comment (kronbichler) [inline source/distributed/tria.cc]: Do you still use periodic_face_pairs_level0 in this function? I can't seem to find it. I guess you only need periodic_face_map here (for which I don't quite understand how you can simply call 'tria.periodic_face_map' below? `enforce_mesh_balance_over_periodic_boundaries` is a separate function, and periodic_face_map is not public?)

File: source/grid/tria_accessor.cc [hunk 9]
```diff
@@ -1869,6 +1869,325 @@ CellAccessor<dim, spacedim>::neighbor_of_coarser_neighbor (const unsigned int ne
...
 
 template <int dim, int spacedim>
+bool
+CellAccessor<dim, spacedim>::has_periodic_neighbor (const unsigned int i_face) const
+{
+  /*
+   * Implementation note: In all of the functions corresponding to periodic faces
+   * we mainly use the Triangulation::periodic_face_map to find the
+   * information about periodically connected faces. So, we actually search in
+   * this std::map and return the cell_face on the other side of periodic boundary.
```
Comment (kronbichler) [inline source/grid/tria_accessor.cc]: of periodic boundary -> of the periodic boundary

File: source/grid/tria_accessor.cc [hunk 10]
```diff
@@ -1869,6 +1869,325 @@ CellAccessor<dim, spacedim>::neighbor_of_coarser_neighbor (const unsigned int ne
...
+   * For this search process, we have three options:
+   *
+   * 1- Using the [] operator of std::map: This option results in a more readalbe
+   *    code, but requires an extra iteration in the map. Becasue when we call [] on
```
Comment (kronbichler) [inline source/grid/tria_accessor.cc]: Becasue -> Because

File: source/grid/tria_accessor.cc [hunk 11]
```diff
@@ -1869,6 +1869,325 @@ CellAccessor<dim, spacedim>::neighbor_of_coarser_neighbor (const unsigned int ne
...
 template <int dim, int spacedim>
+bool
+CellAccessor<dim, spacedim>::has_periodic_neighbor (const unsigned int i_face) const
+{
+  /*
+   * Implementation note: In all of the functions corresponding to periodic faces
+   * we mainly use the Triangulation::periodic_face_map to find the
+   * information about periodically connected faces. So, we actually search in
+   * this std::map and return the cell_face on the other side of periodic boundary.
+   * For this search process, we have three options:
```
Comment (kronbichler) [inline source/grid/tria_accessor.cc]: I guess you mean 2 options here?

File: source/grid/tria_accessor.cc [hunk 12]
```diff
@@ -1869,6 +1869,325 @@ CellAccessor<dim, spacedim>::neighbor_of_coarser_neighbor (const unsigned int ne
...
+  Assert (nb_face_pair != this->tria->periodic_face_map.end(),
+          TriaAccessorExceptions::ExcNoPeriodicNeighbor());
+  cell_iterator p_nb_of_p_nb = nb_face_pair->second.first.first;
+  TriaIterator<TriaAccessor<dim - 1, dim, spacedim> > parent_face_it = p_nb_of_p_nb->face(nb_face_pair->second.first.second);
+  for (unsigned int i_subface = 0; i_subface < parent_face_it->n_children(); ++i_subface)
+    if (parent_face_it->child_index(i_subface) == my_face_index)
+      return (std::pair<unsigned int, unsigned int>(face_num_of_nb, i_subface));
+  /*
+   * Obviously, if the execution reaches to this point, some of our assumptions should have
+   * been false. The most important one is, the user has called this funciton on a face
```
Comment (kronbichler) [inline source/grid/tria_accessor.cc]: funciton -> function

File: source/grid/tria_accessor.cc [hunk 13]
```diff
@@ -1869,6 +1869,325 @@ CellAccessor<dim, spacedim>::neighbor_of_coarser_neighbor (const unsigned int ne
...
+  AssertIndexRange (i_face, GeometryInfo<dim>::faces_per_cell);
+  typedef TriaIterator<CellAccessor<dim, spacedim> > cell_iterator;
+  typedef std::pair<cell_iterator, unsigned int> cell_face_pair;
+  typedef std::pair<cell_face_pair, std::bitset<3> > oriented_cell_face_pair;
+  cell_iterator my_it(*this);
+  typename std::map<const cell_face_pair, oriented_cell_face_pair>::const_iterator my_face_pair =
+    this->tria->periodic_face_map.find(std::pair<cell_iterator, unsigned int>(my_it, i_face));
+  Assert (my_face_pair != this->tria->periodic_face_map.end(),
+          TriaAccessorExceptions::ExcNoPeriodicNeighbor());
+  return my_face_pair->second.first.first->index();
```
Comment (kronbichler) [inline source/grid/tria_accessor.cc]: Isn't the return value of this function simply `periodic_neighbor(i_face)->index()`?

File: source/grid/tria_accessor.cc [hunk 14]
```diff
@@ -1869,6 +1869,325 @@ CellAccessor<dim, spacedim>::neighbor_of_coarser_neighbor (const unsigned int ne
...
+  AssertIndexRange (i_face, GeometryInfo<dim>::faces_per_cell);
+  typedef TriaIterator<CellAccessor<dim, spacedim> > cell_iterator;
+  typedef std::pair<cell_iterator, unsigned int> cell_face_pair;
+  typedef std::pair<cell_face_pair, std::bitset<3> > oriented_cell_face_pair;
+  cell_iterator my_it(*this);
+  typename std::map<const cell_face_pair, oriented_cell_face_pair>::const_iterator my_face_pair =
+    this->tria->periodic_face_map.find(std::pair<cell_iterator, unsigned int>(my_it, i_face));
+  Assert (my_face_pair != this->tria->periodic_face_map.end(),
+          TriaAccessorExceptions::ExcNoPeriodicNeighbor());
+  return my_face_pair->second.first.first->level();
```
Comment (kronbichler) [inline source/grid/tria_accessor.cc]: Similarly to above, this should be `periodic_neighbor(i_face)->level()` if I'm not overlooking something.

File: tests/mpi/periodic_neighbor_01.cc [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+  : refn_cycle(0), mpi_comm(MPI_COMM_WORLD), the_grid(mpi_comm)
+{
+  Assert(dim == 2 || dim == 3, ExcMessage("Only implemented for the 2D and 3D case!"));
+  comm_rank = Utilities::MPI::this_mpi_process(mpi_comm);
+  comm_size = Utilities::MPI::n_mpi_processes(mpi_comm);
+  std::vector<unsigned> repeats(dim, 2);
+  Point<dim> p1, p2, periodic_transfer;
+  if (dim == 2)
+    {
+      p2 = { 16., 16. };
```
Comment (kronbichler) [inline tests/mpi/periodic_neighbor_01.cc]: Isn't using initializer lists a C++11 feature?

File: tests/mpi/periodic_neighbor_01.cc [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+      periodic_transfer = { 0.0, 16. };
+    }
+  if (dim == 3)
+    {
+      p2 = { 16., 16., 16. };
+      periodic_transfer = { 0.0, 16., 0.0 };
+    }
+  GridGenerator::subdivided_hyper_rectangle(the_grid, repeats, p1, p2, true);
+  std::vector<GridTools::PeriodicFacePair<cell_iterator>> periodic_faces;
```
Comment (kronbichler) [inline tests/mpi/periodic_neighbor_01.cc]: Same here, `>>` will not compile on older (non-C++11) compilers -> use `> >`

File: tests/mpi/periodic_neighbor_01.cc [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+  GridTools::collect_periodic_faces(the_grid, 2, 3, 0, periodic_faces, periodic_transfer);
+  the_grid.add_periodicity(periodic_faces);
+}
+
+template <int dim>
+void periodicity_tests<dim>::refine_grid(const unsigned &n)
```
Comment (kronbichler) [inline tests/mpi/periodic_neighbor_01.cc]: You might want to pass n by value, not by reference

File: tests/mpi/periodic_neighbor_01.cc [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+
+using namespace dealii;
+
+template <int dim>
+struct periodicity_tests
+{
+  typedef TriaIterator<CellAccessor<dim>> cell_iterator;
+  typedef typename Triangulation<dim>::active_cell_iterator active_cell_iterator;
+  periodicity_tests();
+  ~periodicity_tests();
```
Comment (kronbichler) [inline tests/mpi/periodic_neighbor_01.cc]: You implement an empty destructor below. Since the class is non-virtual, you might just as well skip the destructor entirely. The compiler will do everything for you.

File: tests/mpi/periodicity_04.cc [hunk 19]
```diff
@@ -230,6 +231,50 @@ void check
...
+      const Point<dim> face_center_2 = cell_2->face(face_no_2)->center();
+      Assert(std::min(std::abs(face_center_1(dim-1)-3.), std::abs(face_center_1(dim-1)+3.))<1.e-8, ExcInternalError());
+      Assert(std::min(std::abs(face_center_2(dim-1)-3.), std::abs(face_center_2(dim-1)+3.))<1.e-8, ExcInternalError());
+      if (cell_1->level()== cell_2->level())
+        for (unsigned int c = 0; c<dim-1; ++c)
+          if (std::abs(face_center_1(c)-face_center_2(c))>1.e-8)
+            {
+              std::cout << "face_center_1: " << face_center_1 << std::endl;
+              std::cout << "face_center_2: " << face_center_2 << std::endl;
+              for (auto it = triangulation.periodic_face_map.begin();
```
Comment (kronbichler) [inline tests/mpi/periodicity_04.cc]: auto is C++11.

File: tests/mpi/periodic_neighbor_01.cc [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+      test_3D.check_periodicity();
+
+      /* Next, we want to check the case of nonuniform mesh */
+      if (Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+        deallog << "Next, we check the periodic faces on an adaptively refined mesh."
+                << std::endl;
+      test_3D.refine_grid(2);
+      //  test_3D.write_grid();
+      test_3D.check_periodicity();
+    }
```
Comment (kronbichler) [inline tests/mpi/periodic_neighbor_01.cc]: Nice test!

File: tests/mpi/periodic_neighbor_01.mpirun=3.output [hunk 21]
```diff
@@ -0,0 +1,343 @@
+DEAL::All of the cells with periodic neighbors on rank 2
```
Comment (kronbichler) [inline tests/mpi/periodic_neighbor_01.mpirun=3.output]: I don't understand why you appear to only get output from processor 2. I guess that you might want to use 'MPILogInitAll' here, see e.g. in `tests/mpi/mg_01.cc` for how to use it.

**Comment by kronbichler** [comment]
In reference to #2212

File: include/deal.II/grid/tria.h [hunk 22]
```diff
@@ -1468,6 +1474,18 @@ class Triangulation : public Subscriptor
...
+   * periodic face pairs on level 0 for later access during the
+   * identification of ghost cells for the multigrid hierarchy and for
+   * setting up the periodic_face_map.
+   */
+  std::vector<GridTools::PeriodicFacePair<cell_iterator> > periodic_face_pairs_level_0;
+
+  /**
+   * If add_periodicity() is called, this variable stores the active periodic face pairs.
+   */
+  std::map<std::pair<cell_iterator, unsigned int>, std::pair<std::pair<cell_iterator, unsigned int>, std::bitset<3> > > periodic_face_map;
```
Comment (kronbichler) [inline include/deal.II/grid/tria.h]: I've been thinking about the name of the data structure here - it looks a bit lengthy with all those pairs (and cell_iterator is a bit wasteful because each item stores a pointer to the triangulation - but it is handy). On the other hand, it is nice to have a clear statement of the involved data structures so I don't see a better solution. Any thoughts?

**Comment by kronbichler** [comment]
In the mailing list, we discussed about an interface of the type `neighbor_or_periodic_neighbor`. I see that you chose a different solution with `periodic_neighbor` only. I don't feel strongly about it and am fine with the proposed solution, but we should discuss whether everyone agrees on the approach to have separate functions and require action in various places, including assemblers for DG elements, in terms of a check of the form
```
if (cell->at_boundary(f) && !cell->has_periodic_neighbor(f))
  // call boundary function
else if (cell->at_boundary(f))
  // access periodic neighbor
else
  // access usual neighbor
```
rather than joining both else cases.

File: tests/mpi/periodicity_04.cc (hunk 19 above)
Comment (masterleinad) [inline tests/mpi/periodicity_04.cc]: You've got me there. :-)

File: include/deal.II/grid/tria.h (hunk 22 above)
Comment (bangerth) [inline include/deal.II/grid/tria.h]: Instead of nesting `std::pair` classes, you could also use `std_cxx11::tuple` (unless the pairing is natural, of course).

File: include/deal.II/grid/tria.h [hunk 23]
```diff
@@ -2964,6 +2982,28 @@ class Triangulation : public Subscriptor
   void load (Archive &ar,
              const unsigned int version);
 
+
+  /**
+    * Declare the (coarse) face pairs given in the argument of this function
+    * as periodic. This way it it possible to obtain neighbors across periodic
+    * boundaries.
+    * This function initializes periodic_face_map that stores the active periodic faces.
```
Comment (bangerth) [inline include/deal.II/grid/tria.h]: I'd say remove this last sentence -- it states something that the user has no relationship with (namely how a private member variable of the class is initialized).

File: include/deal.II/grid/tria_accessor.h [hunk 24]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
   /**
    * @}
    */
+  /**
+   * @name Dealing with periodic neighbor
```
Comment (bangerth) [inline include/deal.II/grid/tria_accessor.h]: neighbor -> neighbor cells? neighbors?

File: include/deal.II/grid/tria_accessor.h [hunk 25]
```diff
@@ -2489,6 +2496,116 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+   * For a cell with a periodic neighbor at its @c ith face, this function
+   * returns the face number of that periodic neighbor such that, the
+   * current cell is the periodic neighbor of that neighbor. In other words
+   * the following assertion holds for those cells which have a periodic
+   * neighbor with the same or a higher level of refinement as the current
+   * cell:
+   * @c {cell->periodic_neighbor(i)->
+   *     periodic_neighbor(cell->periodic_neighbor_of_periodic_neighbor(i))==cell}
+   * For the cells with a coarser periodic neighbor, one should use
+   * @c periodic_neighbor_of_coarser_periodic_neighbor and
```
Comment (bangerth) [inline include/deal.II/grid/tria_accessor.h]: Instead of using `@c` here, just say `periodic_neighbor_of_coarser_periodic_neighbor()` and doxygen will use the parentheses as an indication that this is a function that can be linked to. (It will then also use the appropriate font.)

File: source/distributed/tria.cc [hunk 26]
```diff
@@ -2244,6 +2244,7 @@ namespace parallel
         }
 
       this->update_number_cache ();
+      Triangulation<dim, spacedim>::update_periodic_face_map();
```
Comment (bangerth) [inline source/distributed/tria.cc]: Just say `this->u_p_f_m()`.

File: source/distributed/tria.cc [hunk 27]
```diff
@@ -2774,6 +2773,7 @@ namespace parallel
...
```
Comment (bangerth) [inline source/distributed/tria.cc]: same here

File: source/distributed/tria.cc [hunk 28]
```diff
@@ -3892,8 +3864,8 @@ namespace parallel
 
 
       refinement_in_progress = false;
-
       this->update_number_cache ();
+      Triangulation<dim, spacedim>::update_periodic_face_map();
```
Comment (bangerth) [inline source/distributed/tria.cc]: `this->u_p_f_m()`.

File: source/distributed/tria.cc [hunk 29]
```diff
@@ -3963,6 +3935,7 @@ namespace parallel
 
       // update how many cells, edges, etc, we store locally
       this->update_number_cache ();
+      Triangulation<dim, spacedim>::update_periodic_face_map();
```
Comment (bangerth) [inline source/distributed/tria.cc]: `this->`

File: source/distributed/tria.cc [hunk 30]
```diff
@@ -4960,6 +4831,7 @@ namespace parallel
         }
 
       this->update_number_cache ();
+      Triangulation<dim, spacedim>::update_periodic_face_map();
```
Comment (bangerth) [inline source/distributed/tria.cc]: this->

File: source/grid/tria_accessor.cc [hunk 31]
```diff
@@ -1869,6 +1869,325 @@ CellAccessor<dim, spacedim>::neighbor_of_coarser_neighbor (const unsigned int ne
...
+   *    code, but requires an extra iteration in the map. Becasue when we call [] on
+   *    std::map, with a key which does not exist in the std::map, that key will be
+   *    created and the default value will be returned by []. This is not desirable.
+   *    So, one has to first check if the key exists in the std::map and if it exists,
+   *    then use the [] operator. The existence check is possible using std::map::find()
+   *    or std::map::count(). Using this option will result in two iteration cycles
+   *    through the map. First, existence check, then returning the value.
+   *
+   * 2- Using std::map::find(): This option is less readble, but theoretically
+   *    faster. Because, it results in one iteration through std::map object.
```
Comment (bangerth) [inline source/grid/tria_accessor.cc]: . Because, it -> ", because it"

File: source/distributed/tria.cc (hunk 8 above)
Comment (masterleinad) [inline source/distributed/tria.cc]: Now periodic_face_map is private.

File: include/deal.II/grid/tria.h (hunk 23 above)
Comment (masterleinad) [inline include/deal.II/grid/tria.h]: Done.

File: include/deal.II/grid/tria.h (hunk 22 above)
Comment (masterleinad) [inline include/deal.II/grid/tria.h]: I am fine with using std_cxx11::tuple<std::pair<...>,std::pair<...>,std::bitset<3> > if you like this better.

Comment (kronbichler) [inline include/deal.II/grid/tria.h]: @masterleinad It's a bit difficult to read your code because github has removed something, but I think @bangerth referred to choosing `std::map<std::pair<cell_iterator, unsigned int>, std_cxx11::tuple<cell_iterator, unsigned int, std::bitset<3> > >`, so keeping the pair for the key but using tuple for the value because there you need to put three arguments. I would agree with that choice.

Comment (masterleinad) [inline include/deal.II/grid/tria.h]: Fine with me.

Comment (bangerth) [inline include/deal.II/grid/tria.h]: I updated the markup of @masterleinad 's comment. In the end, I'm agnostic which is better, I just wanted to point out another possibility.

Comment (masterleinad) [inline include/deal.II/grid/tria.h]: Thank you, @bangerth. If no one feels strongly about this, I leave it as it is.

File: source/grid/tria_accessor.cc (hunk 14 above)
Comment (samiiali) [inline source/grid/tria_accessor.cc]: No, you are not overlooking, I implemented these two functions before `periodic_neighbor` !

File: tests/mpi/periodic_neighbor_01.cc (hunk 15 above)
Comment (samiiali) [inline tests/mpi/periodic_neighbor_01.cc]: Do you mean initialization list in line 125, or initializing `p2` by a C style array in line 134? Initialization list in 125 is not C++11. I also believe that in 134, there is an implicit conversion from `{16., 16.}` to a C array which has no issue with pre-C++11 compilers. What am I missing here?

File: tests/mpi/periodic_neighbor_01.cc (hunk 20 above)
Comment (samiiali) [inline tests/mpi/periodic_neighbor_01.cc]: Thanks !

File: tests/mpi/periodic_neighbor_01.cc (hunk 15 above)
Comment (bangerth) [inline tests/mpi/periodic_neighbor_01.cc]: Line 134. It's not a C-style array initialization, but in C++11 is mapped to a copy constructor. This doesn't work in C++98, though. You'll have to write this as
```
  p2[0] = 16.;
  p2[1] = 16.;
```
or
```
  p2 = Point<dim>(16,16);
```
Same in line 139.

Comment (samiiali) [inline tests/mpi/periodic_neighbor_01.cc]: Nice point ! thanks.

File: include/deal.II/grid/tria.h (hunk 22 above)
Comment (kronbichler) [inline include/deal.II/grid/tria.h]: I agree.

**Comment by kronbichler** [comment]
I think that all issues I brought up have been addressed. Any comments regarding the interface that only returns the periodic neighbor rather than `neighbor_or_periodic_neighbor`? I would suggest to wait for a day or two to see if anyone has remarks - if not, I think this is ready for merge. In the meantime, please rebase.

**Comment by masterleinad** [comment]
I am in favor to have `neighbor_or_periodic_neighbor` but like to hear comments by others about this as well.

**Comment by bangerth** [comment]
I think having `neighbor_or_periodic_neighbor()` is the way to go. But that could of course also be done in a separate, follow-up patch.

**Comment by samiiali** [comment]
Actually, in our first implementation we had `neighbor_or_periodic_neighbor`. Then we decided to have all of these nine functions, which means we should have something like `neighbor_or_periodic_neighbor_of_coarser_neighbor_or_periodic_neighbor`. Which was a bit too long for a name. Instead there could be both `neighbor` and `periodic_neighbor` and another set of functions like `generalized_neighbor`. Or, maybe making `periodic_neighbor` private and make `generalized_neighbor` public for users. Since, these functions do not have any legacy users, this decision can be made based on the best design.

**Comment by bangerth** [comment]
I think you made the right call. There probably aren't too many places where one would need `neighbor_or_periodic_neighbor_of_coarser_neighbor_or_periodic_neighbor`, and if one does, it can be emulated by the functions you implemented. But `neighbor_or_periodic_neighbor()` (or `generalized_neighbor() -- I like that name) is something we will need frequently. So I'd say it's useful to have (and trivial to implement based on the functions you already have), even if there are no equivalent shortcuts for the other functions that allow querying neighborship.

In fact, I've also thought about that we should go through at least parts of the code and see where we should use `neighbor_or_periodic_neighbor()`. One is in the mesh smoothing algorithms in `tria.cc`. Others are in those places where we build sparsity patterns, for example.

But one step at a time.

**Comment by kronbichler** [comment]
I agree on providing all functionality in terms of `periodic_neighbor_xxx` as is done right now, but adding `neighbor or_periodic_neighbor()` as an additional function that can get used at the interesting places in subsequent PRs (the mesh smoothing algorithms for example was something we discussed in #2116 - we might be able to get rid of the code introduced in that PR).

**Comment by samiiali** [comment]
As a final confirmation, the name of this additional function should be `generalized_neighbor`or `neighbor_or_periodic_neighbor`. I like the first one better, but the downside is the name `generalized_neighbor` is not as descriptive as `neighbor_or_periodic_neighbor`. So, as soon as we agree on the name, I can push this addition through this pull request.

**Comment by bangerth** [comment]
Let's go with the more descriptive name -- we've usually gone down that road in deal.II, and I think it's served us reasonably well over the years.

**Comment by kronbichler** [comment]
I agree.

**Comment by masterleinad** [comment]
Let's use `neighbor_or_periodic_neighbor`.

File: include/deal.II/grid/tria_accessor.h [hunk 32]
```diff
@@ -2489,6 +2496,126 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+   */
+  unsigned int
+  periodic_neighbor_of_periodic_neighbor (const unsigned int i) const;
+
+  /**
+   * If a cell has periodic neighbor, this function returns the face number
+   * of the neighbor, which is connected to this cell.
+   */
+  unsigned int
+  periodic_neighbor_face_no (const unsigned int i) const;
```
Comment (tjhei) [inline include/deal.II/grid/tria_accessor.h]: what is the argument i here?

Comment (tjhei) [inline include/deal.II/grid/tria_accessor.h]: I also don't understand the difference between ``periodic_neighbor_face_no`` and ``periodic_neighbor_of_periodic_neighbor``

File: include/deal.II/grid/tria_accessor.h [hunk 33]
```diff
@@ -2489,6 +2496,126 @@ class CellAccessor :  public TriaAccessor<dim,dim,spacedim>
...
+   */
+  std::pair<unsigned int, unsigned int>
+  periodic_neighbor_of_coarser_periodic_neighbor (const unsigned face_no) const;
+
+  /**
+   * This function returns the index of the periodic neighbor. If there is
+   * no periodic neighbor at the given face, the returned value is -1.
+   */
+  int
+  periodic_neighbor_index (const unsigned int i) const;
```
Comment (tjhei) [inline include/deal.II/grid/tria_accessor.h]: Is the return value the cell index (``cell->index()``) or what is meant by "the index"?

File: include/deal.II/grid/tria_accessor.h (hunk 32 above)
Comment (samiiali) [inline include/deal.II/grid/tria_accessor.h]: Concerning `i` as the argument of `periodic_neighbor_face_no`, I have fixed the documentation in the following way:
```
If a cell has periodic neighbor at its @c ith face, this function
returns the face number of the periodic neighbor, which is
connected to the @c ith face of this cell.
 ```
Concerning the difference between `periodic_neighbor_face_no` and `periodic_neighbor_of_periodic_neighbor`, I checked the difference between `neighbor_face_no` and `neighbor_of_neighbor`. It looks to me that when the neighbor is coarser `neighbor_of_neighbor` throws an exception, while `neighbor_face_no` returns the first element of `neighbor_of_coarser_neighbor`. However, in the `Triangulation::periodic_face_map` Daniel stores all of the coarser periodic neighbors of each cell (Daniel can correct me if I am wrong). In this case, `periodic_neighbor_face_no` and `periodic_neighbor_of_periodic_neighbor` are the same. If I am wrong (i.e. we do not have all of the coarser neighbors of a cell in `Triangulation::periodic_face_map`), then I need to fix this issue.

File: include/deal.II/grid/tria_accessor.h (hunk 33 above)
Comment (samiiali) [inline include/deal.II/grid/tria_accessor.h]: Yes, the index means the same thing as `cell->index()`. Since, this word is used in the documentation of the `neighbor_index()` function, I assume the users should understand what we mean by that. I fixed the documentation in the following way, though:
```
This function returns the index of the periodic neighbor at the
@c ith face of the current cell. If there is no periodic neighbor
at the given face, the returned value is -1.
```

File: include/deal.II/grid/tria_accessor.h (hunk 32 above)
Comment (samiiali) [inline include/deal.II/grid/tria_accessor.h]: I updated my previous comment.

**Comment by masterleinad** [comment]
Is there anything else that should be addressed?

**Comment by kronbichler** [comment]
I think this looks very good now. Thanks for the effort @samiiali and @masterleinad.

**Comment by kronbichler** [comment]
I get the following warning:
```
source/grid/tria_accessor.cc:1947:1: warning: control reaches end of non-void function [-Wreturn-type]
```
Would you mind opening a follow-up PR that fixes this by returning a dummy argument?

**Comment by masterleinad** [comment]
Sure, see PR #2240
